R/load_cell_graphs.R

```diff
@@ -214,6 +262,142 @@ LoadCellGraphs.CellGraphAssay <- function (
+  add_marker_counts = TRUE,
+  force = FALSE,
+  chunk_size = 10,
+  cl = NULL,
+  verbose = TRUE,
+  ...
+) {
+
+  # Validate input parameters
+  stopifnot(
+    "'cells' must be a non-empty character vector of cell names" =
+      is.character(cells) &&
+      (length(cells) > 0)
+  )
+  stopifnot(
+    "'cells' must be present in 'object'" =
+      all(cells %in% colnames(object))
+  )
+  load_as <- match.arg(load_as, choices = c("bipartite", "Anode", "linegraph"))
+
+  # Check if cells are already loaded
+  loaded_graphs <- !sapply(slot(object, name = "cellgraphs")[cells], is.null)
+  if ((sum(!loaded_graphs) == 0) & !force) {
+    if (verbose && check_global_verbosity())
+      cli_alert_info("All cells are alredy loaded. Returning CellGraphAssay unmodified.")
+    return(object)
+  } else {
+    if (force) {
+      slot(object, "cellgraphs")[cells] <- rep(list(NULL), length(cells)) %>% setNames(nm = cells)
+      loaded_graphs <- !sapply(slot(object, name = "cellgraphs")[cells], is.null)
+    }
+    cells_to_load <- setdiff(cells, cells[loaded_graphs])
+    if (verbose && check_global_verbosity() & (length(cells_to_load) < length(cells)))
+      cli_alert_info(glue("{length(cells) - length(cells_to_load)} CellGraphs loaded.",
+                          " Loading remaining {length(cells_to_load)} CellGraphs."))
+  }
+
+  # Find where components are located
+  fs_map_unnested_filtered <- slot(object, name = "fs_map") %>%
+    tidyr::unnest(cols = "id_map") %>%
+    filter(current_id %in% cells)
+  fs_map_nested_filtered <- fs_map_unnested_filtered %>%
+    tidyr::nest(data = c("current_id", "original_id"))
+
+  # Load cell graphs in chunks
+  cg_list_full <- lapply(seq_len(nrow(fs_map_nested_filtered)), function(i) {
+    f <- fs_map_nested_filtered$pxl_file[i]
+
+    # Unzip the edgelist parquet file to tmpdir
+    unzip(f, exdir = tempdir(), files = "edgelist.parquet")
+    unz_pq_file <- file.path(tempdir(), "edgelist.parquet")
+    pq_file <- fs::file_temp(ext = "parquet")
+
+    # The file is renamed to make sure it has a unique path
+    fs::file_move(unz_pq_file, pq_file)
+
+    # Read the edgelist in memory, but only the necessary columns
+    ar <- arrow::read_parquet(pq_file,
+                              col_select = c("upia", "upib", "marker", "component"),
+                              as_data_frame = FALSE)
+
+    if (verbose && check_global_verbosity()) {
+      cli_alert(glue("   Loading CellGraphs for {nrow(fs_map_nested_filtered$data[[i]])} cells ",
+                     "from sample {fs_map_nested_filtered$sample[i]}"))
+    }
+
+    # Split data into chunks determined by chunk_size
+    id_data_chunks <- fs_map_nested_filtered$data[[i]] %>%
+      mutate(group = ceiling(seq_len(n())/chunk_size)) %>%
+      rename(component = current_id) %>%
+      group_by(group) %>%
+      group_split()
+
+    # Read cellgraphs
+    cg_list <- pbapply::pblapply(id_data_chunks, function(id_chunk) {
```

Comment:
Does pbapply parallelize successfully across platforms now? It used to have issues with that on Windows.

Reply:
I have set the the `cl` parameter to NULL which defaults to serial processing. Users can pass an integer to decide on the number of cores to use which should work on Mac OS but is ignored on Windows. On windows, you can instead create a cluster and pass it with cl. To it's up to the user to decide how to handle parallelization.